// locale-loader/lib/generateXlfData/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

exports.default = generateXlfData;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _xmlJs = require('xml-js');

var _xmlJs2 = _interopRequireDefault(_xmlJs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function generateXlfData(_ref) {
  var rawData = _ref.rawData,
      sourceLocale = _ref.sourceLocale,
      supportedLocales = _ref.supportedLocales,
      sourceFolder = _ref.sourceFolder,
      exportType = _ref.exportType;

  var isFull = exportType.toLowerCase() === 'full';
  var onlyTranslated = exportType.toLowerCase() === 'translated';
  var jsonData = {};
  var allLocales = supportedLocales.filter(function (locale) {
    return locale !== sourceLocale;
  });

  allLocales.forEach(function (locale) {
    jsonData[locale] = {
      _declaration: {
        _attributes: {
          version: '1.0'
        }
      },
      xliff: {
        _attributes: {
          version: '1.2',
          xmlns: 'urn:oasis:names:tc:xliff:document:1.2'
        }
      }
    };
  });

  (0, _keys2.default)(rawData).forEach(function (folderPath) {
    var folderData = rawData[folderPath];
    var sourceFile = folderData.files[sourceLocale];
    if (sourceFile) {
      var keys = (0, _keys2.default)(sourceFile.data);
      supportedLocales.forEach(function (locale) {
        if (locale !== sourceLocale) {
          var targetFile = folderData.files[locale];
          var fileName = targetFile && targetFile.file || locale + '.js';
          var original = _path2.default.relative(sourceFolder, _path2.default.join(folderData.path, fileName));
          var transUnits = [];
          keys.forEach(function (key) {
            if (onlyTranslated) {
              if (targetFile && targetFile.data[key] && (!targetFile.data[key].source || targetFile.data[key].source === sourceFile.data[key].value)) {
                var unit = {
                  _attributes: {
                    id: '[' + key + ']'
                  },
                  source: {
                    _text: sourceFile.data[key].value
                  },
                  target: {
                    _text: targetFile.data[key].value
                  }
                };
                transUnits.push(unit);
              }
            } else {
              var diff = !targetFile || !targetFile.data[key] || targetFile.data[key].source && targetFile.data[key].source !== sourceFile.data[key].value;
              if (!onlyTranslated && diff || isFull) {
                var _unit = {
                  _attributes: {
                    id: '[' + key + ']'
                  },
                  source: {
                    _text: sourceFile.data[key].value
                  },
                  target: {
                    _text: diff ? sourceFile.data[key].value : targetFile.data[key].value
                  }
                };
                transUnits.push(_unit);
              }
            }
          });
          if (transUnits.length) {
            var data = {
              _attributes: {
                original: original,
                'source-language': sourceLocale,
                'target-language': locale,
                datatype: 'plaintext'
              },
              body: {
                'trans-unit': transUnits
              }
            };
            if (!jsonData[locale].xliff.file) {
              jsonData[locale].xliff.file = [];
            }
            jsonData[locale].xliff.file.push(data);
          }
        }
      });
    }
  });
  var xlfData = {};
  allLocales.forEach(function (locale) {
    xlfData[locale] = _xmlJs2.default.json2xml(jsonData[locale], { compact: true, spaces: 4 });
  });
  return xlfData;
}
